File: models/partidos.py
"""CRUD de pichangas (partidos programados)."""

from database.connection import get_connection


def crear_partido(fecha, hora, cancha, cupo_max, costo_cancha, costo_por_jugador, notas=""):
    conexion = get_connection()
    try:
        cursor = conexion.execute(
            """
            INSERT INTO partidos (fecha, hora, cancha, cupo_max, costo_cancha, costo_por_jugador, notas)
            VALUES (?, ?, ?, ?, ?, ?, ?)
            """,
            (fecha, hora, cancha.strip(), cupo_max, costo_cancha, costo_por_jugador, notas.strip()),
        )
        conexion.commit()
        return cursor.lastrowid
    finally:
        conexion.close()


def obtener_partido(partido_id):
    conexion = get_connection()
    try:
        fila = conexion.execute("SELECT * FROM partidos WHERE id = ?", (partido_id,)).fetchone()
        return dict(fila) if fila else None
    finally:
        conexion.close()


def listar_partidos(estado=None):
    conexion = get_connection()
    try:
        consulta = "SELECT * FROM partidos"
        parametros = ()
        if estado:
            consulta += " WHERE estado = ?"
            parametros = (estado,)
        consulta += " ORDER BY fecha, hora"
        filas = conexion.execute(consulta, parametros).fetchall()
        return [dict(f) for f in filas]
    finally:
        conexion.close()


def actualizar_partido(partido_id, fecha, hora, cancha, cupo_max, costo_cancha, costo_por_jugador, notas=""):
    conexion = get_connection()
    try:
        conexion.execute(
            """
            UPDATE partidos
            SET fecha = ?, hora = ?, cancha = ?, cupo_max = ?, costo_cancha = ?,
                costo_por_jugador = ?, notas = ?
            WHERE id = ?
            """,
            (fecha, hora, cancha.strip(), cupo_max, costo_cancha, costo_por_jugador, notas.strip(), partido_id),
        )
        conexion.commit()
    finally:
        conexion.close()


def cambiar_estado(partido_id, estado):
    conexion = get_connection()
    try:
        conexion.execute("UPDATE partidos SET estado = ? WHERE id = ?", (estado, partido_id))
        conexion.commit()
    finally:
        conexion.close()


def duplicar_partido(partido_id, nueva_fecha, nueva_hora):
    """Repite un partido (misma cancha/cupo/costo) en una fecha/hora nueva."""
    original = obtener_partido(partido_id)
    return crear_partido(
        nueva_fecha,
        nueva_hora,
        original["cancha"],
        original["cupo_max"],
        original["costo_cancha"],
        original["costo_por_jugador"],
        original["notas"] or "",
    )

File: web/src/actions/partidos.ts
"use server";

import { revalidatePath } from "next/cache";
import { prisma } from "@/lib/prisma";
import { requireAdmin } from "@/lib/require-auth";

export async function crearPartido(formData: FormData) {
  await requireAdmin();

  const formato = String(formData.get("formato") ?? "futbol7");
  await prisma.partido.create({
    data: {
      fecha: String(formData.get("fecha")),
      hora: String(formData.get("hora")),
      cancha: String(formData.get("cancha")).trim(),
      cupoMax: Number(formData.get("cupoMax")),
      costoCancha: Number(formData.get("costoCancha")),
      costoPorJugador: Number(formData.get("costoPorJugador")),
      notas: String(formData.get("notas") ?? "").trim() || null,
      formato: formato === "futbol11" ? "futbol11" : "futbol7",
      numEquipos: Math.min(3, Math.max(2, Number(formData.get("numEquipos")) || 2)),
      arquerosMin: Math.max(0, Number(formData.get("arquerosMin")) || 0),
      arquerosMax: Math.max(1, Number(formData.get("arquerosMax")) || 2),
    },
  });

  revalidatePath("/dashboard/partidos");
}

export async function cambiarEstadoPartido(partidoId: number, estado: string) {
  await requireAdmin();
  await prisma.partido.update({ where: { id: partidoId }, data: { estado } });
  revalidatePath("/dashboard/partidos");
}

export async function duplicarPartido(partidoId: number, nuevaFecha: string, nuevaHora: string) {
  await requireAdmin();
  const original = await prisma.partido.findUniqueOrThrow({ where: { id: partidoId } });
  await prisma.partido.create({
    data: {
      fecha: nuevaFecha,
      hora: nuevaHora,
      cancha: original.cancha,
      cupoMax: original.cupoMax,
      costoCancha: original.costoCancha,
      costoPorJugador: original.costoPorJugador,
      notas: original.notas,
      formato: original.formato,
      numEquipos: original.numEquipos,
      arquerosMin: original.arquerosMin,
      arquerosMax: original.arquerosMax,
    },
  });
  revalidatePath("/dashboard/partidos");
}
